lib/hdb/common.c

```diff
@@ -153,6 +154,23 @@ _hdb_fetch_kvno(krb5_context context, HDB *db, krb5_const_principal principal,
 	    krb5_data_free(&value);
 	    return ret;
 	}
+
+	if ((flags & HDB_F_FOR_AS_REQ) && (flags & HDB_F_CANON) == 0) {
+	    krb5_principal tmp;
+
+	    /*
+	     * Return the name the client asked for. Should be done in the KDC,
+	     * but the interactions with the Samba backend would be complex, so
+	     * do it here.
+	     */
+	    ret = krb5_copy_principal(context, principal, &tmp);
+	    if (ret) {
+		krb5_data_free(&value);
+		return ret;
+	    }
+	    krb5_free_principal(context, entry->entry.principal);
```

Comment:
Do the `krb5_free_principal(context, entry->entry.principal)` first then the `krb5_copy_principal(context, principal, &entry->entry.principal)`, then you can do without a `tmp` variable.

Reply:
Yes, but if krb5_copy_principal fails, then we've killed the principal. Which I reckon smells funny, and may impact logging too I suppose.